src/main/kotlin/com/autonomousapps/extension/SourceSetsHandler.kt

```diff
@@ -0,0 +1,41 @@
+package com.autonomousapps.extension
+
+import org.gradle.api.Named
+import org.gradle.api.NamedDomainObjectFactory
+import org.gradle.api.model.ObjectFactory
+import javax.inject.Inject
+
+/**
+ * TODO: kdoc.
+ */
+open class SourceSetsHandler @Inject constructor(
+  private val sourceSetName: String,
+  projectPath: String,
+  objects: ObjectFactory
+) : Named {
+
+  override fun getName(): String = sourceSetName
+
+  internal val project = objects.newInstance(ProjectIssueHandler::class.java, projectPath)
+
+  /**
+   * Gets the specified issue out of [project], and sets [sourceSetName] on it for later filtering by
+   * [FilterAdviceTask][com.autonomousapps.tasks.FilterAdviceTask].
+   */
+  internal fun issueOf(mapper: (ProjectIssueHandler) -> Issue): Issue {
+    return mapper(project).apply {
+      sourceSet.set(sourceSetName)
+    }
+  }
+
+  internal class Factory(
```

Comment:
Not sure I actually need a custom factory, but thinking about how Gradle instantiates custom objects via reflection was giving me a headache.

Reply:
Generally, I would make all types to be instantiated by Gradle `abstract` instead of `open`. Small practical advantage is that you won't accidentally instantiate them without Gradle's factories.

Then oftentimes you won't need `objects.*()` calls. But you can have `abstract val` that Gradle fills as we do in the tasks here.

However, this automatic stuff does not work with _domain object containers_ and custom types in all cases. I am also not sure when and why. I tend to avoid _domain object containers_ as these are so overloaded in their API (right now). I think there is also no clear statement from Gradle on this. For usability in a DSL, I prefer a clear interface with as little methods as possible. But that's a a different fundamental/philosophical discussion. :)

Anyway, just some background which might already know.

I think in this case, you have to do it like this to pass the `projectPath` to all objects. I don't know of a better way to do it.